### PR TITLE
cmd/tailscale/cli: [set] handle selectively modifying routes/exit node

### PR DESCRIPTION
**Pull request by maisem**
Noticed this while debugging something else, we would reset all routes if either `--advertise-exit-node` or `--advertise-routes` were set. This handles correctly updating them.

Also added tests.

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Comment by DentonGentry** [comment]
Fixing this means that starting the exit node with `tailscale up --advertise-exit-node --advertise-routes=172.16.0.0/16` will result in other being able to use the exit node and also reach 172.16.0.0 through a local interface on the exit node?

**Comment by maisem** [comment]
> Fixing this means that starting the exit node with `tailscale up --advertise-exit-node --advertise-routes=172.16.0.0/16` will result in other being able to use the exit node and also reach 172.16.0.0 through a local interface on the exit node?

This is unrelated to `tailscale up`, its for the new `tailscale set` subcommand added in 19b558657308a0cd6d8e2eac272737552fb04725.

File: cmd/tailscale/cli/set.go [hunk 1]
```diff
@@ -128,3 +136,27 @@ func runSet(ctx context.Context, args []string) (retErr error) {
 	_, err = localClient.EditPrefs(ctx, maskedPrefs)
 	return err
 }
+
+func calcAdvertiseRoutesForSet(advertiseExitNodeSet, advertiseRoutesSet bool, curPrefs *ipn.Prefs, setArgs setArgsT) (routes []netip.Prefix, err error) {
```
Comment (bradfitz) [inline cmd/tailscale/cli/set.go:140]: this signature could use some docs, even if it's unexported

File: cmd/tailscale/cli/set_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import (
+	"net/netip"
+	"reflect"
+	"testing"
+
+	"tailscale.com/ipn"
+	"tailscale.com/net/tsaddr"
+)
+
+func ptrTo[T any](v T) *T { return &v }
```
Comment (bradfitz) [inline cmd/tailscale/cli/set_test.go:16]: hah, we have this everywhere :)

Comment (bradfitz) [reply cmd/tailscale/cli/set_test.go:16]: (we'll clean 'em all later)

File: cmd/tailscale/cli/set_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+func TestCalcAdvertiseRoutesForSet(t *testing.T) {
+	pfx := netip.MustParsePrefix
+	tests := []struct {
+		name     string
+		exit     *bool
+		routes   *string
+		existing []netip.Prefix
```
Comment (bradfitz) [inline cmd/tailscale/cli/set_test.go:24]: "had" or "was" is a bit shorter? don't care, though

File: cmd/tailscale/cli/set_test.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+func ptrTo[T any](v T) *T { return &v }
+
+func TestCalcAdvertiseRoutesForSet(t *testing.T) {
+	pfx := netip.MustParsePrefix
+	tests := []struct {
+		name     string
+		exit     *bool
+		routes   *string
```
Comment (bradfitz) [inline cmd/tailscale/cli/set_test.go:23]: setExit, setRoutes ?

**Review by bradfitz — APPROVED** [review]
Nice tests!

File: cmd/tailscale/cli/set.go (hunk 1 above)
Comment (maisem) [reply cmd/tailscale/cli/set.go:146]: Done

File: cmd/tailscale/cli/set_test.go (hunk 3 above)
Comment (maisem) [reply cmd/tailscale/cli/set_test.go]: Done

File: cmd/tailscale/cli/set_test.go (hunk 4 above)
Comment (maisem) [reply cmd/tailscale/cli/set_test.go]: Done

File: cmd/tailscale/cli/set_test.go (hunk 2 above)
Comment (maisem) [reply cmd/tailscale/cli/set_test.go:16]: its a very useful func!

**Pull request merged**